src/bootstrap/step.rs

```diff
@@ -838,14 +838,19 @@ invalid rule dependency graph detected, was a rule added and maybe typo'd?
             } else {
                 &self.build.config.target
             };
-            // If --target was specified but --host wasn't specified, don't run
-            // any host-only tests
+            // Determine the actual targets participating in this rule.
+            // NOTE: We should keep the full projection from build triple to
+            // the hosts for the dist steps, now that the hosts array above is
+            // truncated to avoid duplication of work in that case. Therefore
+            // the original non-shadowed hosts array is used below.
             let arr = if rule.host {
+                // If --target was specified but --host wasn't specified,
+                // don't run any host-only tests
                 if self.build.flags.target.len() > 0 &&
                    self.build.flags.host.len() == 0 {
-                    &hosts[..0]
+                    &[]
                 } else {
-                    hosts
+                    &self.build.config.host[..]
```

Comment:
And and as the final thing I think here we'll want the same logic where if `--host` is specified then we use that here. For that it should mean that we just check `if self.build.flags.host.len() != 0` then we use that array here (no matter) what. You could probably structure this as:

```rust
if self.build.flags.host.len() > 0 {
    &self.build.flags.host[..]
} else if self.build.flags.target.len() > 0 {
    &[]
} else {
    &self.build.config.host[..]
}
```